File: supporter-product-data/src/main/scala/com/gu/lambdas/AddSupporterRatePlanItemToQueueLambda.scala
package com.gu.lambdas

import com.amazonaws.services.lambda.runtime.Context
import com.gu.lambdas.AddSupporterRatePlanItemToQueueLambda.addToQueue
import com.gu.model.StageConstructors
import com.gu.model.dynamo.SupporterRatePlanItemCodecs._
import com.gu.model.states.AddSupporterRatePlanItemToQueueState
import com.gu.services.{AlarmService, ConfigService, S3Service, SqsService}
import com.gu.supporterdata.model.{Stage, SupporterRatePlanItem}
import com.typesafe.scalalogging.StrictLogging
import kantan.csv._
import kantan.csv.ops._

import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Future

trait TimeOutCheck {
  def timeRemainingMillis: Int
}

class ContextTimeOutCheck(context: Context) extends TimeOutCheck {
  override def timeRemainingMillis = context.getRemainingTimeInMillis
}

class AddSupporterRatePlanItemToQueueLambda
    extends Handler[AddSupporterRatePlanItemToQueueState, AddSupporterRatePlanItemToQueueState] {
  override protected def handlerFuture(input: AddSupporterRatePlanItemToQueueState, context: Context) = {
    addToQueue(StageConstructors.fromEnvironment, input, new ContextTimeOutCheck(context))
  }
}

object AddSupporterRatePlanItemToQueueLambda extends StrictLogging {
  val maxBatchSize = 5
  val timeoutBufferInMillis = maxBatchSize * 5 * 1000

  def addToQueue(
      stage: Stage,
      state: AddSupporterRatePlanItemToQueueState,
      timeOutCheck: TimeOutCheck,
  ): Future[AddSupporterRatePlanItemToQueueState] = {
    logger.info(s"Starting to add subscriptions to queue for ${state.recordCount} records from ${state.filename}")

    val s3Object = S3Service.streamFromS3(stage, state.filename)
    val csvReader = s3Object.getObjectContent.asCsvReader[SupporterRatePlanItem](rfc.withHeader)
    val sqsService = SqsService(stage)
    val alarmService = AlarmService(stage)

    if (csvReader.isEmpty) {
      s3Object.close()
      alarmAndExit(alarmService, s"The specified CSV file ${state.filename} was empty")
    }

    val unProcessed = getUnprocessedItems(csvReader, state.processedCount)

    // Close this after retrieving all the items
    s3Object.close()

    val validUnprocessed = unProcessed.collect { case (Right(item), index) => (item, index) }
    val invalidUnprocessedIndexes = unProcessed.collect { case (Left(_), index) => index }

    if (invalidUnprocessedIndexes.nonEmpty) {
      alarmAndExit(
        alarmService,
        s"there were ${invalidUnprocessedIndexes.length} read failures from file ${state.filename} with line numbers ${invalidUnprocessedIndexes
            .mkString(",")}",
      )
    }

    val batches = validUnprocessed.grouped(10).toList

    val processedCount = writeBatchesUntilTimeout(
      state.processedCount,
      batches,
      timeOutCheck,
      sqsService,
      alarmService,
    )

    val maybeSaveSuccessTime =
      if (processedCount == state.recordCount)
        ConfigService(stage).putLastSuccessfulQueryTime(state.attemptedQueryTime)
      else Future.successful(())

    maybeSaveSuccessTime.map(_ => state.copy(processedCount = processedCount))
  }

  def alarmAndExit(alarmService: AlarmService, message: String) = {
    logger.error(
      s"CSV read failure: $message",
    )
    alarmService.triggerCsvReadAlarm
    throw new RuntimeException(message)
  }

  def getUnprocessedItems(csvReader: CsvReader[ReadResult[SupporterRatePlanItem]], processedCount: Int) =
    csvReader.zipWithIndex.drop(processedCount).toList

  def writeBatchesUntilTimeout(
      processedCount: Int,
      batches: List[List[(SupporterRatePlanItem, Int)]],
      timeOutCheck: TimeOutCheck,
      sqsService: SqsService,
      alarmService: AlarmService,
  ): Int =
    batches.foldLeft(processedCount) { (processedCount, batch) =>
      if (timeOutCheck.timeRemainingMillis < timeoutBufferInMillis) {
        logger.info(
          s"Aborting processing - time remaining: ${timeOutCheck.timeRemainingMillis / 1000} seconds, buffer: ${timeoutBufferInMillis / 1000} seconds",
        )
        return processedCount
      }
      logger.info(
        s"Continuing processing with batch of ${batch.length} - time remaining: ${timeOutCheck.timeRemainingMillis / 1000} seconds, buffer: ${timeoutBufferInMillis / 1000} seconds",
      )
      sqsService.sendBatch(batch)

      val (_, highestProcessedIndex) = batch.last
      val newProcessedCount = highestProcessedIndex + 1
      logger.info(s"$newProcessedCount items processed")
      newProcessedCount
    }
}

File: supporter-product-data-ts/src/typescript/lambdas/addSupporterRatePlanItemToQueueLambda.ts
import type { Context, Handler } from "aws-lambda";
import { stageFromEnvironment } from "../model/stage";
import type { SupporterRatePlanItem } from "../model/supporterRatePlanItem";
import { supporterRatePlanItemFromCsvRow } from "../model/supporterRatePlanItem";
import { AlarmService } from "../services/alarmService";
import { ConfigService } from "../services/configService";
import { parseCsvStreamWithHeader } from "../services/csvService";
import { S3Service } from "../services/s3Service";
import { SqsService } from "../services/sqsService";
import type { AddSupporterRatePlanItemToQueueState } from "./types";

const maxBatchSize = 5;
const timeoutBufferInMillis = maxBatchSize * 5 * 1000;

type IndexedItem = [SupporterRatePlanItem, number];

interface AddToQueueDeps {
  streamCsvRows: (
    filename: string
  ) => AsyncIterable<Record<string, string>> | Iterable<Record<string, string>>;
  sendBatch: (batch: IndexedItem[]) => Promise<void>;
  triggerCsvReadAlarm: () => Promise<void>;
  triggerSqsWriteAlarm: () => Promise<void>;
  putLastSuccessfulQueryTime: (time: string) => Promise<void>;
}

const buildDeps = (): AddToQueueDeps => {
  const stage = stageFromEnvironment();
  const s3Service = new S3Service();
  const queueUrl = process.env.QUEUE_URL;
  if (queueUrl === undefined) {
    throw new Error("QUEUE_URL must be set");
  }

  const sqsService = new SqsService(queueUrl);
  const alarmService = new AlarmService(stage);
  const configService = new ConfigService(stage);

  return {
    streamCsvRows: (filename) =>
      parseCsvStreamWithHeader(s3Service.streamObjectLines(stage, filename)),
    sendBatch: (batch) => sqsService.sendBatch(batch),
    triggerCsvReadAlarm: () => alarmService.triggerCsvReadAlarm(),
    triggerSqsWriteAlarm: () => alarmService.triggerSqsWriteAlarm(),
    putLastSuccessfulQueryTime: (time) =>
      configService.putLastSuccessfulQueryTime(time),
  };
};

export const addToQueue = async (
  state: AddSupporterRatePlanItemToQueueState,
  getRemainingTimeInMillis: () => number,
  deps: AddToQueueDeps
): Promise<AddSupporterRatePlanItemToQueueState> => {
  console.info("Starting to add subscriptions to queue", {
    filename: state.filename,
    recordCount: state.recordCount,
    processedCount: state.processedCount,
    remainingMillis: getRemainingTimeInMillis(),
  });

  let rowIndex = 0;
  let processedCount = state.processedCount;
  let seenAnyRow = false;
  let batch: IndexedItem[] = [];

  const flushBatch = async (): Promise<void> => {
    if (batch.length === 0) {
      return;
    }

    if (getRemainingTimeInMillis() < timeoutBufferInMillis) {
      console.info("Aborting processing due to remaining lambda time", {
        remainingMillis: getRemainingTimeInMillis(),
        timeoutBufferInMillis,
        processedCount,
      });
      // Return without flushing — the step function will re-invoke with the
      // current processedCount so we resume from where we left off
      batch = [];
      return;
    }

    try {
      await deps.sendBatch(batch);
      console.info("Successfully wrote SQS batch", {
        batchSize: batch.length,
        firstIndex: batch[0]?.[1],
        lastIndex: batch[batch.length - 1]?.[1],
      });
    } catch (error) {
      console.error("Failed to write SQS batch", error);
      await deps.triggerSqsWriteAlarm();
    }

    const highestIndex = batch[batch.length - 1]?.[1];
    if (highestIndex !== undefined) {
      processedCount = highestIndex + 1;
    }
    batch = [];
  };

  for await (const row of deps.streamCsvRows(state.filename)) {
    seenAnyRow = true;

    // Skip rows already processed in a previous invocation
    if (rowIndex < state.processedCount) {
      rowIndex += 1;
      continue;
    }

    // Bail out early if we're running out of time — the step function will
    // re-invoke with the updated processedCount
    if (getRemainingTimeInMillis() < timeoutBufferInMillis) {
      console.info("Aborting processing due to remaining lambda time", {
        remainingMillis: getRemainingTimeInMillis(),
        timeoutBufferInMillis,
        processedCount,
      });
      break;
    }

    let item: SupporterRatePlanItem;
    try {
      item = supporterRatePlanItemFromCsvRow(row, rowIndex + 2);
    } catch (error) {
      console.warn("Failed to decode CSV row", { rowIndex, error });
      await deps.triggerCsvReadAlarm();
      throw new Error(
        `Failed to decode CSV row at index ${rowIndex} in file ${
          state.filename
        }: ${String(error)}`
      );
    }

    batch.push([item, rowIndex]);

    if (batch.length >= maxBatchSize) {
      await flushBatch();
    }

    rowIndex += 1;
  }

  // Flush any remaining items in the last partial batch
  await flushBatch();

  if (!seenAnyRow) {
    await deps.triggerCsvReadAlarm();
    throw new Error(`The specified CSV file ${state.filename} was empty`);
  }

  console.info("Finished writing to SQS", {
    filename: state.filename,
    processedCount,
    recordCount: state.recordCount,
    complete: processedCount === state.recordCount,
  });

  if (processedCount === state.recordCount) {
    console.info("All records processed, updating lastSuccessfulQueryTime", {
      attemptedQueryTime: state.attemptedQueryTime,
    });
    await deps.putLastSuccessfulQueryTime(state.attemptedQueryTime);
  }

  return {
    ...state,
    processedCount,
  };
};

const fromContext =
  (context: Context): (() => number) =>
  () =>
    context.getRemainingTimeInMillis();

export const handler: Handler<
  AddSupporterRatePlanItemToQueueState,
  AddSupporterRatePlanItemToQueueState
> = async (state, context) =>
  addToQueue(state, fromContext(context), buildDeps());
